pvlib/tests/iotools/test_psm3.py

```diff
@@ -49,48 +61,72 @@ def assert_psm3_equal(header, data, expected):
     assert (data.index.tzinfo.zone == 'Etc/GMT%+d' % -header['Time Zone'])
 
 
+def delay_get_psm3(*args, **kwargs):
+    """Repeat get_psm3() call if OVER_RATE_LIMIT error occurs otherwise
+    allow HTTPError to raise. Sleep times increase by one second for
+    every OVER_RATE_LIMIT HTTPError in a row. Return (header, data) from
+    get_psm3() as normal.
+    """
+    sleep_time = 0
+    while True:
+        try:
+            # execute get_psm3
+            header, data = psm3.get_psm3(*args, **kwargs)
+        except HTTPError as e:
+            if "OVER_RATE_LIMIT" in str(e):
+                # retry command after sleeping for a time
+                time.sleep(sleep_time)
+                sleep_time += 1
+                continue
+            # raise HTTPError if not due to OVER_RATE_LIMIT
+            raise
+        else:
+            break
+    return header, data
+
+
 @needs_pandas_0_22
 def test_get_psm3_tmy():
     """test get_psm3 with a TMY"""
-    header, data = psm3.get_psm3(LATITUDE, LONGITUDE, DEMO_KEY, PVLIB_EMAIL,
+    header, data = delay_get_psm3(LATITUDE, LONGITUDE, DEMO_KEY, PVLIB_EMAIL,
                                  names='tmy-2017')
     expected = pd.read_csv(TMY_TEST_DATA)
     assert_psm3_equal(header, data, expected)
     # check errors
     with pytest.raises(HTTPError):
         # HTTP 403 forbidden because api_key is rejected
-        psm3.get_psm3(LATITUDE, LONGITUDE, api_key='BAD', email=PVLIB_EMAIL)
+        delay_get_psm3(LATITUDE, LONGITUDE, api_key='BAD', email=PVLIB_EMAIL)
     with pytest.raises(HTTPError):
         # coordinates were not found in the NSRDB
-        psm3.get_psm3(51, -5, DEMO_KEY, PVLIB_EMAIL)
+        delay_get_psm3(51, -5, DEMO_KEY, PVLIB_EMAIL)
     with pytest.raises(HTTPError):
         # names is not one of the available options
-        psm3.get_psm3(LATITUDE, LONGITUDE, DEMO_KEY, PVLIB_EMAIL, names='bad')
+        delay_get_psm3(LATITUDE, LONGITUDE, DEMO_KEY, PVLIB_EMAIL, names='bad')
 
 
 @needs_pandas_0_22
 def test_get_psm3_singleyear():
     """test get_psm3 with a single year"""
-    header, data = psm3.get_psm3(LATITUDE, LONGITUDE, DEMO_KEY, PVLIB_EMAIL,
+    header, data = delay_get_psm3(LATITUDE, LONGITUDE, DEMO_KEY, PVLIB_EMAIL,
                                  names='2017', interval=30)
     expected = pd.read_csv(YEAR_TEST_DATA)
     assert_psm3_equal(header, data, expected)
     # check leap day
-    _, data_2012 = psm3.get_psm3(LATITUDE, LONGITUDE, DEMO_KEY, PVLIB_EMAIL,
+    _, data_2012 = delay_get_psm3(LATITUDE, LONGITUDE, DEMO_KEY, PVLIB_EMAIL,
                                  names='2012', interval=60, leap_day=True)
     assert len(data_2012) == (8760+24)
     # check errors
     with pytest.raises(HTTPError):
```

Comment:
why are these error checks essentially repeated from `test_get_psm3_tmy`? can we safely delete them and save some of our api calls?

Reply:
I had the same thought in discussing with @kanderso-nrel in #870. The error catching seems to test NREL's API more than get_psm3() but I do appreciate the thoroughness of the tests for injecting "bad" inputs and demonstrating that they fail.  I'm not attached to it either way.

Reply:
It seems to me that these failures can safely be expected for both the tmy and non-tmy requests. Would be even better if they were in their own test function.

Reply:
I was just going to suggest that the tests be split up so we can use the `@pytest.mark.flaky(reruns=5, reruns_delay=2)` decorator over each API call. We still have to check if the error contains "OVER_RATE_LIMIT" and fail the test since that too is an HTTPError.

Reply:
I'm okay with dumping these api tests, sorry for adding them initially

Reply:
Not at all @mikofski, I appreciate the thoroughness. 

There could be great use in these as potential early warnings of potentially breaking changes to the API that could warn us by giving a clearer picture of what is failing. This could backfire if there's a lot of API changes that create a nuisance but that by itself could be a warning as well. 

I'm sure this type of issue has been raised and resolved before so I'm ok either way. The newest PR consolidates all error tests into a single parametrized test so that way help keep things tidy enough to move forward.